crates/datagen/src/lib.rs

```diff
@@ -337,194 +355,393 @@ impl InputGenerator {
+                    eoi = true;
+                    running = false;
+                    consumer.eoi();
+                }
+            } else {
+                std::thread::park();
+            }
+        }
+
+        Ok(())
     }
 
-    #[allow(clippy::too_many_arguments, clippy::type_complexity)]
+    #[allow(clippy::too_many_arguments)]
     async fn worker_thread(
-        worker_idx: usize,
+        work_receiver: AsyncReceiver<Work>,
+        completion_sender: UnboundedSender<Completion>,
         config: DatagenInputConfig,
-        shared_state: Arc<
-            Vec<(
-                AtomicUsize,
-                RateLimiter<NotKeyed, InMemoryState, DefaultClock, NoOpMiddleware>,
-            )>,
-        >,
         schema: Relation,
         consumer: Box<dyn InputConsumer>,
         mut parser: Box<dyn Parser>,
-        queue: Arc<InputQueue>,
-        notifier: Arc<Notify>,
-        status: Arc<Atomic<PipelineState>>,
-        generated: Arc<AtomicUsize>,
+        rate_limiters: Arc<Vec<RateLimiter<NotKeyed, InMemoryState, DefaultClock, NoOpMiddleware>>>,
+        datagen_thread: Thread,
     ) {
         let mut buffer = Vec::new();
+        let mut buffer_start = 0;
         static START_ARR: &[u8; 1] = b"[";
         static END_ARR: &[u8; 1] = b"]";
         static REC_DELIM: &[u8; 1] = b",";
 
-        for (plan, (progress, rate_limiter)) in config.plan.into_iter().zip(shared_state.iter()) {
-            // Inserting in batches improves performance by around ~80k records/s for
-            // me, so we always do it rather than giving the user the option.
-            // If we have a low rate, it's necessary to adjust the batch-size down otherwise no inserts might be visible
-            // for a long time (until a batch is full).
-            let batch_size: usize = min(plan.rate.unwrap_or(u32::MAX) as usize, 10_000);
-            let per_thread_chunk: usize = plan.worker_chunk_size.unwrap_or(batch_size);
+        while let Ok(work) = work_receiver.recv().await {
+            datagen_thread.unpark();
 
-            let limit = plan.limit.unwrap_or(usize::MAX);
+            let Work {
+                batch: Batch { plan_idx, rows },
+                rate_limit,
+                seed,
+            } = work;
+            let plan = &config.plan[plan_idx];
+            let batch_size: usize = min(plan.rate.unwrap_or(u32::MAX) as usize, 10_000);
             let schema = schema.clone();
-            let mut generator = RecordGenerator::new(worker_idx, config.seed, plan, schema);
+            let mut generator = RecordGenerator::new(seed, plan, &schema);
 
             // Count how long we took to so far to create a batch
             // If we end up taking too long we send a batch earlier even if we don't reach `batch_size`
-            const BATCH_CREATION_TIMEOUT: TokioDuration = TokioDuration::from_secs(1);
+            const BATCH_CREATION_TIMEOUT: StdDuration = StdDuration::from_secs(1);
```

Comment:
is there a deeper reason this needs to be threaded now?

Reply:
> is there a deeper reason this needs to be threaded now?

The most straightforward way I could think of to make datagen fault-tolerant was to introduce a single thread that dispatched work to worker tasks. It honestly hadn't occurred to me that the dispatcher could be a task instead of a thread. I'm happy to try to figure out whether that's an option if you'd like it.

By the way, the change highlighted above is just code churn: `tokio::time::Duration` is just a re-export of `std::time::Duration` and I didn't like how it was being called out as apparently tokio-specific.